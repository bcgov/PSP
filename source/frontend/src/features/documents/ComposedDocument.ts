import { DocumentRelationshipType } from 'constants/documentRelationshipType';
import {
  Api_Document,
  Api_DocumentMetadataUpdate,
  Api_DocumentRelationship,
  Api_DocumentType,
  Api_DocumentUpdateRequest,
  Api_DocumentUploadRequest,
} from 'models/api/Document';
import {
  Api_Storage_DocumentDetail,
  Api_Storage_DocumentMetadata,
  Api_Storage_DocumentTypeMetadataType,
} from 'models/api/DocumentStorage';
import Api_TypeCode from 'models/api/TypeCode';

export interface ComposedDocument {
  mayanMetadata?: Api_Storage_DocumentMetadata[];
  pimsDocument?: Api_Document;
  documentDetail?: Api_Storage_DocumentDetail;
  mayanFileId?: number;
}

export class DocumentRow {
  id?: number;
  mayanDocumentId?: number;
  documentType?: Api_DocumentType;
  statusTypeCode?: Api_TypeCode<string>;
  fileName?: string;
  isFileAvailable?: boolean;
  appCreateTimestamp?: string;
  appCreateUserid?: string;
  relationshipId?: number;
  relationshipType?: DocumentRelationshipType;
  parentId?: number;

  public static fromApi(relationship: Api_DocumentRelationship): DocumentRow {
    const row: DocumentRow = new DocumentRow();
    row.id = relationship.document?.id;
    row.documentType = relationship.document?.documentType;
    row.mayanDocumentId = relationship.document?.mayanDocumentId;
    row.statusTypeCode = relationship.document?.statusTypeCode;
    row.fileName = relationship.document?.fileName;
    row.appCreateTimestamp = relationship.document?.appCreateTimestamp;
    row.appCreateUserid = relationship.document?.appCreateUserid;

    row.relationshipId = relationship.id;
    row.relationshipType = relationship.relationshipType;
    row.parentId = relationship.parentId;
    return row;
  }

  public static toApi(document: DocumentRow): Api_DocumentRelationship {
    return {
      id: document.relationshipId,
      relationshipType: document.relationshipType,
      parentId: document.parentId,
      document: {
        id: document.id,
        mayanDocumentId: document.mayanDocumentId,
        documentType: document.documentType,
        statusTypeCode: document.statusTypeCode,
        fileName: document.fileName,
      },
    };
  }
}

export class DocumentUploadFormData {
  public documentTypeId: string;
  public documentStatusCode: string;
  public documentMetadata: Record<string, string>;

  public constructor(
    initialStatus: string,
    documentType: string,
    metadata: Api_Storage_DocumentTypeMetadataType[],
  ) {
    this.documentStatusCode = initialStatus;
    this.documentTypeId = documentType;
    this.documentMetadata = {};

    metadata.forEach(metaType => {
      this.documentMetadata[metaType.metadata_type?.id?.toString() || '-'] = '';
    });
  }

  public toRequestApi(file: File, documentType: Api_DocumentType): Api_DocumentUploadRequest {
    var metadata: Api_DocumentMetadataUpdate[] = [];
    for (const key in this.documentMetadata) {
      const value = this.documentMetadata[key];
      metadata.push({
        metadataTypeId: Number(key),
        value: value,
      });
    }

    return {
      documentType: documentType,
      documentStatusCode: this.documentStatusCode,
      file: file,
      documentMetadata: metadata,
    };
  }
}

export class DocumentUpdateFormData {
  public documentId: number;
  public mayanDocumentId: number;
  public documentStatusCode: string;
  public documentMetadata: Record<string, string>;

  public static fromApi(
    composedDocument: ComposedDocument,
    metadataTypes: Api_Storage_DocumentTypeMetadataType[],
  ): DocumentUpdateFormData {
    var model = new DocumentUpdateFormData();
    model.documentId = composedDocument.pimsDocument?.id || 0;
    model.mayanDocumentId = composedDocument.pimsDocument?.mayanDocumentId || 0;
    model.documentStatusCode = composedDocument.pimsDocument?.statusTypeCode?.id?.toString() || '';
    model.documentMetadata = {};
    metadataTypes.forEach(metaType => {
      var foundMetadata = composedDocument.mayanMetadata?.find(
        currentMeta => currentMeta.metadata_type.id === metaType.metadata_type?.id,
      );
      model.documentMetadata[metaType.metadata_type?.id?.toString() || '-'] =
        foundMetadata?.value ?? '';
    });
    return model;
  }

  public toRequestApi(): Api_DocumentUpdateRequest {
    var metadata: Api_DocumentMetadataUpdate[] = [];

    for (const key in this.documentMetadata) {
      const value = this.documentMetadata[key];
      const metadataTypeId = Number(key);
      metadata.push({
        value: value,
        metadataTypeId: metadataTypeId,
      });
    }

    return {
      documentId: this.documentId,
      mayanDocumentId: this.mayanDocumentId,
      documentStatusCode: this.documentStatusCode,
      documentMetadata: metadata,
    };
  }

  private constructor() {
    this.documentId = -1;
    this.mayanDocumentId = -1;
    this.documentStatusCode = '';
    this.documentMetadata = {};
  }
}
