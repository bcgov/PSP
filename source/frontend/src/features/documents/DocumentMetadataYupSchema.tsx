import * as Yup from 'yup';

import { ApiGen_Mayan_DocumentTypeMetadataType } from '@/models/api/generated/ApiGen_Mayan_DocumentTypeMetadataType';

export const getDocumentMetadataYupSchema = (
  mayanMetadata: ApiGen_Mayan_DocumentTypeMetadataType[],
) => {
  const metadataSchema: Record<string, Yup.StringSchema> = {};
  for (const data of mayanMetadata) {
    if (data.metadata_type?.id && data.required) {
      metadataSchema[`${data.metadata_type?.id}`] = Yup.string().required(
        `${data.metadata_type?.label} is required`,
      );
    }
  }

  return Yup.object().shape({
    documentMetadata: Yup.object().shape(metadataSchema),
    documentTypeId: Yup.string().required('Document type is required'),
  });
};
