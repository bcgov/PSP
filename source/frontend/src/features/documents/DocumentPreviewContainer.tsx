import moment from 'moment';
import { FC, useCallback, useContext, useEffect, useState } from 'react';

import LoadingBackdrop from '@/components/common/LoadingBackdrop';
import { ModalContext } from '@/contexts/modalContext';
import { ApiGen_Mayan_FilePage } from '@/models/api/generated/ApiGen_Mayan_FilePage';
import { ApiGen_System_HttpStatusCode } from '@/models/api/generated/ApiGen_System_HttpStatusCode';

import { DocumentViewerContext } from './context/DocumentViewerContext';
import { IDocumentPreviewViewProps } from './DocumentPreviewView';
import { createFileDownload } from './DownloadDocumentButton';
import { useDocumentProvider } from './hooks/useDocumentProvider';

export interface LoadedPage {
  loadedDocumentImageDataUrl?: string;
  error?: string;
  mayanPage: ApiGen_Mayan_FilePage;
}

export interface IDocumentPreviewContainerProps {
  View: React.FC<IDocumentPreviewViewProps>;
  createFileUrl?: (obj: Blob | MediaSource) => string;
}

export const DocumentPreviewContainer: FC<
  React.PropsWithChildren<IDocumentPreviewContainerProps>
> = ({ View, createFileUrl = URL.createObjectURL }) => {
  const { showDocumentPreview, previewDocumentId, setShowDocumentPreview, setPreviewDocumentId } =
    useContext(DocumentViewerContext);
  const { setModalContent, setDisplayModal } = useContext(ModalContext);
  const {
    retrieveDocumentDetail,
    downloadDocumentFilePageImage,
    getDocumentFilePageList,
    downloadWrappedDocumentFileLatest,
    retrieveDocumentDetailLoading,
    getDocumentFilePageListLoading,
  } = useDocumentProvider();
  const [documentPages, setDocumentPages] = useState<LoadedPage[]>([]);
  const [currentPage, setCurrentPage] = useState<number>(0);

  //Load a document page from mayan as a dataUrl (that can be supplied to an <img>)
  const loadDocumentPageImageUrl = useCallback(
    async (
      documentId: number,
      documentFileId: number,
      documentFilePageId: number,
    ): Promise<string> => {
      //todo: handle image failure
      return createFileUrl(
        await downloadDocumentFilePageImage(documentId, documentFileId, documentFilePageId),
      );
    },
    [downloadDocumentFilePageImage, createFileUrl],
  );

  const resetDocumentPreview = useCallback(() => {
    setShowDocumentPreview(false);
    setPreviewDocumentId(null);
    setCurrentPage(0);
    setDocumentPages([]);
  }, [setPreviewDocumentId, setShowDocumentPreview]);

  const loadPagesForDocumentId = useCallback(async () => {
    const data = await retrieveDocumentDetail(previewDocumentId);
    if (
      data?.payload?.datetime_created &&
      moment(data.payload.datetime_created).add(1, 'minutes').isAfter(moment())
    ) {
      // Use case for document preview is a user trying to understand what the contents of a document are.
      // A user should not need to know the contents of a document they just uploaded. Also mayan likely need at least a minute to process the majority of the pages.
      setModalContent({
        title: 'Document Still Processing',
        message:
          'This document has been uploaded very recently, please wait a few minutes before trying to preview it. You may download the document in the meantime (using the "eye" icon).',
        variant: 'info',
        handleOk: () => {
          setDisplayModal(false);
          resetDocumentPreview();
        },
      });
      setDisplayModal(true);
    } else if (
      data?.httpStatusCode !== ApiGen_System_HttpStatusCode.OK ||
      !data?.payload?.file_latest
    ) {
      setModalContent({
        title: 'Document Preview Error',
        message:
          'Failed to retrieve document details. This may indicate the document is still being processed, or failed to upload. If the issue persists, please contact the system administrator.',
        variant: 'error',
        handleOk: () => {
          setDisplayModal(false);
          resetDocumentPreview();
        },
      });
      setDisplayModal(true);
    } else if (data?.payload?.file_latest) {
      const pages = await getDocumentFilePageList(previewDocumentId, data.payload.file_latest.id);

      //todo: handle various doc failures.
      if (pages?.length > 0) {
        const documentPages = pages.map(page => ({ mayanPage: page } as LoadedPage));
        setDocumentPages(documentPages);
      } else {
        setModalContent({
          title: 'Document Preview Error',
          message:
            'Unable to retrieve preview pages for this document. Confirm this document is in a valid format for file preview (document, image, pdf). If you believe that this is a valid file type, and the issue persists, please contact the system administrator.',
          variant: 'error',
          handleOk: () => {
            setDisplayModal(false);
            resetDocumentPreview();
          },
        });
        setDisplayModal(true);
      }
    }
  }, [
    getDocumentFilePageList,
    previewDocumentId,
    resetDocumentPreview,
    retrieveDocumentDetail,
    setDisplayModal,
    setModalContent,
  ]);

  useEffect(() => {
    if (previewDocumentId) {
      loadPagesForDocumentId();
    }
  }, [previewDocumentId, loadPagesForDocumentId]);

  useEffect(() => {
    const page = documentPages[currentPage];
    // lazy load the next page, but do not replace lazy loaded data.
    if (previewDocumentId && page && !page.loadedDocumentImageDataUrl) {
      loadDocumentPageImageUrl(
        previewDocumentId,
        page.mayanPage.document_file_id,
        page.mayanPage.id,
      )
        .then(imageUrl => {
          page.loadedDocumentImageDataUrl = imageUrl;
        })
        .catch(e => {
          console.error(e);
          page.error = e.message;
        });
    }
  }, [currentPage, documentPages, loadDocumentPageImageUrl, previewDocumentId]);
  return (
    <>
      <LoadingBackdrop
        show={retrieveDocumentDetailLoading || getDocumentFilePageListLoading}
        parentScreen
      />
      <View
        currentPage={currentPage}
        setCurrentPage={setCurrentPage}
        showDocumentPreview={showDocumentPreview}
        pages={documentPages}
        resetDocumentPreview={resetDocumentPreview}
        handleDownload={() => {
          downloadWrappedDocumentFileLatest(previewDocumentId).then(file =>
            createFileDownload(file),
          );
        }}
      />
    </>
  );
};
