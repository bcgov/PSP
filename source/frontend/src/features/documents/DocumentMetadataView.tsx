import { Input } from 'components/common/form';
import { SectionField } from 'features/mapSideBar/tabs/SectionField';
import { FieldArray, FormikProps } from 'formik';
import { Api_Storage_DocumentTypeMetadataType } from 'models/api/DocumentStorage';

import { StyledNoData } from './commonStyles';
import { DocumentUpdateFormData, DocumentUploadFormData } from './ComposedDocument';

export interface IDocumentMetadataViewProps {
  mayanMetadata: Api_Storage_DocumentTypeMetadataType[];
  formikProps: FormikProps<DocumentUploadFormData> | FormikProps<DocumentUpdateFormData>;
}

export const DocumentMetadataView: React.FunctionComponent<
  React.PropsWithChildren<IDocumentMetadataViewProps>
> = props => {
  return (
    <>
      {props.mayanMetadata.length === 0 && <StyledNoData>No additional data</StyledNoData>}

      <FieldArray
        name="documentMetadata"
        render={() => (
          <>
            {props.mayanMetadata.map(meta => (
              <SectionField
                labelWidth="4"
                key={`document-metadata-${meta.metadata_type?.name}`}
                label={meta.metadata_type?.label || ''}
                required={meta.required === true}
              >
                <Input
                  data-testid={`metadata-input-${meta.metadata_type?.name}` || ''}
                  field={`documentMetadata.${meta.metadata_type?.id}`}
                  required={meta.required === true}
                  value={props.formikProps.values.documentMetadata[meta.metadata_type?.id || '']}
                />
              </SectionField>
            ))}

            <div style={{ color: 'red' }}>
              {Object.values(props.formikProps.errors).length > 0 && (
                <>Mandatory fields are required.</>
              )}
            </div>
          </>
        )}
      />
    </>
  );
};
