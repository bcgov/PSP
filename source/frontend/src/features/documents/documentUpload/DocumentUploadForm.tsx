import { Button } from 'components/common/buttons/Button';
import { Select, SelectOption } from 'components/common/form';
import TooltipIcon from 'components/common/TooltipIcon';
import LoadingBackdrop from 'components/maps/leaflet/LoadingBackdrop/LoadingBackdrop';
import * as API from 'constants/API';
import { SectionField } from 'features/mapSideBar/tabs/SectionField';
import { Formik, FormikProps } from 'formik';
import useLookupCodeHelpers from 'hooks/useLookupCodeHelpers';
import { Api_DocumentType, Api_DocumentUploadRequest } from 'models/api/Document';
import { Api_Storage_DocumentTypeMetadataType } from 'models/api/DocumentStorage';
import { ChangeEvent, useEffect, useState } from 'react';
import { Col, Row } from 'react-bootstrap';

import { StyledGreySection, StyledH2, StyledH3, StyledScrollable } from '../commonStyles';
import { DocumentUploadFormData } from '../ComposedDocument';
import { DocumentMetadataView } from '../DocumentMetadataView';
import { getDocumentMetadataYupSchema } from '../DocumentMetadataYupSchema';
import { StyledContainer } from '../list/styles';

interface IDocumentUploadFormProps {
  isLoading: boolean;
  formikRef: React.RefObject<FormikProps<DocumentUploadFormData>>;
  initialDocumentType: string;
  documentTypes: Api_DocumentType[];
  mayanMetadataTypes: Api_Storage_DocumentTypeMetadataType[];
  onDocumentTypeChange: (changeEvent: ChangeEvent<HTMLInputElement>) => void;
  onUploadDocument: (uploadRequest: Api_DocumentUploadRequest) => void;
  onCancel: () => void;
}

/**
 * Component that provides functionality to see document information. Can be embedded as a widget.
 */
const DocumentUploadForm: React.FunctionComponent<IDocumentUploadFormProps> = props => {
  const [selectedFile, setSelectedFile] = useState<File | null>(null);
  const documentTypes = props.documentTypes.map<SelectOption>(x => {
    return { label: x.documentType || '', value: x.id?.toString() || '' };
  });

  const { getOptionsByType } = useLookupCodeHelpers();
  const documentStatusTypes = getOptionsByType(API.DOCUMENT_STATUS_TYPES);

  const handleFileInput = (changeEvent: ChangeEvent<HTMLInputElement>) => {
    // handle validations
    if (changeEvent.target !== null) {
      var target = changeEvent.target;
      if (target.files !== null) {
        setSelectedFile(target.files[0]);
        // forces formik to flag the change as dirty
        props.formikRef.current?.setFieldValue('fileSet', true);
      }
    }
  };

  const initialFormData = new DocumentUploadFormData(
    documentStatusTypes[0]?.value?.toString(),
    props.initialDocumentType,
    props.mayanMetadataTypes,
  );
  useEffect(() => {
    const isTypeDirty =
      props.documentTypes.find(x => x.id?.toString() === props.initialDocumentType) !== undefined;
    if (isTypeDirty) {
      // forces formik to flag the change as dirty
      props.formikRef.current?.setFieldValue('isDocumentTypeChanged', isTypeDirty);
    }
  }, [props.formikRef, props.documentTypes, props.initialDocumentType, props.mayanMetadataTypes]);

  return (
    <StyledContainer>
      <LoadingBackdrop show={props.isLoading} />
      <Formik<DocumentUploadFormData>
        innerRef={props.formikRef}
        enableReinitialize
        initialValues={initialFormData}
        validateOnMount={true}
        validationSchema={getDocumentMetadataYupSchema(props.mayanMetadataTypes)}
        onSubmit={async (values: DocumentUploadFormData, { setSubmitting }) => {
          if (selectedFile !== null) {
            const selectedDocumentType = props.documentTypes.find(
              x => x.id === Number(values.documentTypeId),
            );
            if (selectedDocumentType !== undefined) {
              var request = values.toRequestApi(selectedFile, selectedDocumentType);

              await props.onUploadDocument(request);
              setSubmitting(false);
            } else {
              console.error('Selected document type is not valid');
            }
          }
        }}
      >
        {formikProps => (
          <>
            <div className="pb-4">
              Choose the document type and select "Browse" to choose the file to upload from your
              computer or network to PIMS.
            </div>
            <SectionField label="Document type" labelWidth="3" className="pb-2">
              <Select
                data-testid="document-type"
                placeholder="Select Document type"
                field="documentTypeId"
                options={documentTypes}
                onChange={props.onDocumentTypeChange}
              />
            </SectionField>
            <SectionField label="Choose document to upload" labelWidth="12" className="mb-4">
              <div className="pt-2">
                <input
                  data-testid="upload-input"
                  id="uploadInput"
                  type="file"
                  name="documentFile"
                  onChange={handleFileInput}
                />
              </div>
            </SectionField>
            <StyledGreySection>
              <Row className="pb-3">
                <Col xs="auto">
                  <StyledH2>Document information</StyledH2>
                </Col>
                <Col xs="auto">
                  <TooltipIcon
                    toolTipId="initiator-tooltip"
                    toolTip="Information you provide here will be searchable"
                  />
                </Col>
              </Row>
              <SectionField label="Status" labelWidth="4">
                <Select field="documentStatusCode" options={documentStatusTypes} />
              </SectionField>

              <StyledH3>Details</StyledH3>
              <StyledScrollable>
                <DocumentMetadataView
                  mayanMetadata={props.mayanMetadataTypes}
                  formikProps={formikProps}
                ></DocumentMetadataView>
              </StyledScrollable>
            </StyledGreySection>
            <Row className="justify-content-end pt-4">
              <Col xs="auto">
                <Button
                  data-testid="cancel"
                  variant="secondary"
                  type="button"
                  onClick={props.onCancel}
                >
                  Cancel
                </Button>
              </Col>
              <Col xs="auto">
                <Button
                  data-testid="save"
                  type="submit"
                  onClick={formikProps.submitForm}
                  disabled={selectedFile === null}
                >
                  Save
                </Button>
              </Col>
            </Row>
          </>
        )}
      </Formik>
    </StyledContainer>
  );
};

export default DocumentUploadForm;
