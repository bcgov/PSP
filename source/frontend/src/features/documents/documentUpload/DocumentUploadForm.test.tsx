import { createMemoryHistory } from 'history';

import { SelectOption } from '@/components/common/form';
import { mockDocumentTypesResponse } from '@/mocks/documents.mock';
import { mockLookups } from '@/mocks/lookups.mock';
import { ApiGen_Mayan_DocumentTypeMetadataType } from '@/models/api/generated/ApiGen_Mayan_DocumentTypeMetadataType';
import { lookupCodesSlice } from '@/store/slices/lookupCodes';
import { fireEvent, render, RenderOptions, userEvent, waitFor } from '@/utils/test-utils';

import { DocumentUploadFormData } from '../ComposedDocument';
import DocumentUploadForm from './DocumentUploadForm';

const history = createMemoryHistory();

const submitForm = jest.fn();
const handleSubmit = jest.fn();

const handleCancelClick = jest.fn();
const onUploadDocument = jest.fn();
const onDocumentTypeChange = jest.fn();
const onDocumentSelected = jest.fn();

const documentStatusOptions: SelectOption[] = [
  { label: '', value: 'NONE' },
  { label: '', value: 'DRAFT' },
  { label: '', value: 'APPROVD' },
  { label: '', value: 'SIGND' },
  { label: '', value: 'FINAL' },
  { label: '', value: 'AMENDD' },
  { label: '', value: 'CNCLD' },
];

const documentTypeMetadataType: ApiGen_Mayan_DocumentTypeMetadataType[] = [
  {
    id: 1,
    document_type: {
      id: 1,
      label: 'BC Assessment Search',
      delete_time_period: null,
      delete_time_unit: null,
      trash_time_period: null,
      trash_time_unit: null,
    },
    metadata_type: {
      default: '',
      id: 1,
      label: 'Tag',
      lookup: '',
      name: 'Tag',
      parser: '',
      parser_arguments: '',
      url: '',
      validation: '',
      validation_arguments: '',
    },
    required: true,
    url: null,
  },
];

describe('DocumentUploadView component', () => {
  // render component under test
  const setup = (renderOptions: RenderOptions & { initialValues: DocumentUploadFormData }) => {
    const utils = render(
      <DocumentUploadForm
        documentTypes={mockDocumentTypesResponse()}
        isLoading={false}
        mayanMetadataTypes={documentTypeMetadataType}
        onDocumentTypeChange={onDocumentTypeChange}
        onDocumentSelected={onDocumentSelected}
        onUploadDocument={onUploadDocument}
        onCancel={handleCancelClick}
        initialDocumentType={'AMMEND'}
        formikRef={{ current: { submitForm, dirty: true } } as any}
        documentStatusOptions={documentStatusOptions}
      />,
      {
        ...renderOptions,
        store: {
          [lookupCodesSlice.name]: { lookupCodes: mockLookups },
        },
        history,
      },
    );

    return {
      ...utils,
      getSaveButton: () => utils.getByText(/Save/i),
      getCancelButton: () => utils.getByText(/Cancel/i),
    };
  };

  let initialValues: DocumentUploadFormData;
  let file: File;
  beforeEach(() => {
    initialValues = new DocumentUploadFormData(
      'AMEND',
      'BC Assessment Search',
      documentTypeMetadataType,
    );
    initialValues.documentTypeId = '1';
    file = new File(['(⌐□_□)'], 'test.png', { type: 'image/png' });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders as expected', () => {
    setup({ initialValues });
    expect(document.body).toMatchSnapshot();
  });

  it('renders the field', () => {
    const { getByTestId } = setup({ initialValues });
    const textarea = getByTestId('document-type');

    expect(textarea).toBeVisible();
  });

  it('displays input for metadata types', async () => {
    const { getByTestId } = setup({ initialValues });
    const textarea = await getByTestId('metadata-input-Tag');

    expect(textarea).toBeVisible();
  });

  it.skip('should submit form when Submit button is clicked', async () => {
    const { getByTestId } = setup({ initialValues });

    const save = await getByTestId('save');
    // get the upload button
    const uploader = getByTestId('upload-input');

    // simulate upload event and wait until finish
    await waitFor(() =>
      fireEvent.change(uploader, {
        target: { files: [file] },
      }),
    );
    userEvent.click(save);

    expect(handleSubmit).toBeCalled();
    expect(handleCancelClick).not.toBeCalled();
  });
});
