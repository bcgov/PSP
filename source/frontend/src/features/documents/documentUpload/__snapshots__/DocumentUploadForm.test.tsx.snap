// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentUploadView component renders as expected 1`] = `
.c8.btn {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0.4rem 1.2rem;
  min-height: 3rem;
  border: 0.2rem solid transparent;
  border-radius: 0.4rem;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 1.8rem;
  font-family: 'BCSans','Noto Sans',Verdana,Arial,sans-serif;
  font-weight: 700;
  -webkit-letter-spacing: 0.1rem;
  -moz-letter-spacing: 0.1rem;
  -ms-letter-spacing: 0.1rem;
  letter-spacing: 0.1rem;
  cursor: pointer;
}

.c8.btn:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  opacity: 0.8;
}

.c8.btn:focus {
  outline-width: 0.4rem;
  outline-style: solid;
  outline-offset: 1px;
  box-shadow: none;
}

.c8.btn.btn-primary {
  border: none;
}

.c8.btn.btn-secondary {
  background: none;
}

.c8.btn.btn-info {
  border: none;
  background: none;
  padding-left: 0.6rem;
  padding-right: 0.6rem;
}

.c8.btn.btn-info:hover,
.c8.btn.btn-info:active,
.c8.btn.btn-info:focus {
  background: none;
}

.c8.btn.btn-light {
  border: none;
}

.c8.btn.btn-dark {
  border: none;
}

.c8.btn.btn-link {
  font-size: 1.6rem;
  font-weight: 400;
  background: none;
  border: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  min-height: 2.5rem;
  line-height: 3rem;
  -webkit-box-pack: left;
  -webkit-justify-content: left;
  -ms-flex-pack: left;
  justify-content: left;
  -webkit-letter-spacing: unset;
  -moz-letter-spacing: unset;
  -ms-letter-spacing: unset;
  letter-spacing: unset;
  text-align: left;
  padding: 0;
}

.c8.btn.btn-link:hover,
.c8.btn.btn-link:active,
.c8.btn.btn-link:focus {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  border: none;
  background: none;
  box-shadow: none;
  outline: none;
}

.c8.btn.btn-link:disabled,
.c8.btn.btn-link.disabled {
  background: none;
  pointer-events: none;
}

.c8.btn:disabled,
.c8.btn:disabled:hover {
  box-shadow: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  pointer-events: none;
  cursor: not-allowed;
  opacity: 0.65;
}

.c8.Button .Button__icon {
  margin-right: 1.6rem;
}

.c8.Button--icon-only:focus {
  outline: none;
}

.c8.Button--icon-only .Button__icon {
  margin-right: 0;
}

.c2.required::before {
  content: '*';
  position: absolute;
  top: 0.75rem;
  left: 0rem;
}

.c1 {
  font-weight: bold;
}

.c6 {
  width: 100%;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  overflow-y: auto;
}

.c3 {
  padding: 1rem;
}

.c4 {
  font-weight: 700;
}

.c5 {
  font-weight: 700;
  font-size: 1.7rem;
  margin-bottom: 1rem;
  text-align: left;
  padding-top: 1rem;
  border-bottom: solid 0.1rem;
}

.c7 {
  overflow-x: hidden;
  max-height: 50rem;
}

.c0 {
  padding: 1rem;
}

<body>
  <div>
    <div
      class="Toastify"
    />
    <div />
    <div
      class="c0"
    >
      <div
        class="pb-4"
      >
        Choose the document type and select "Browse" to choose the file to upload from your computer or network to PIMS.
      </div>
      <div
        class="pb-2 row"
      >
        <div
          class="pr-0 text-left col-3"
        >
          <label
            class="c1"
          >
            Document type
            :
          </label>
        </div>
        <div
          class="c2 text-left col"
        >
          <div
            class="form-group"
          >
            <select
              class="form-select form-control"
              data-testid="document-type"
              id="input-documentTypeId"
              name="documentTypeId"
            >
              <option
                value=""
              >
                Select Document type
              </option>
              <option
                class="option"
                value="1"
              >
                BC Assessment Search
              </option>
              <option
                class="option"
                value="2"
              >
                Privy Council
              </option>
            </select>
          </div>
        </div>
      </div>
      <div
        class="mb-4 row"
      >
        <div
          class="pr-0 text-left col-12"
        >
          <label
            class="c1"
          >
            Choose document to upload
            :
          </label>
        </div>
        <div
          class="c2 text-left col"
        >
          <div
            class="pt-2"
          >
            <input
              data-testid="upload-input"
              id="uploadInput"
              name="documentFile"
              type="file"
            />
          </div>
        </div>
      </div>
      <div
        class="c3"
      >
        <div
          class="pb-3 row"
        >
          <div
            class="col-auto"
          >
            <h2
              class="c4"
            >
              Document information
            </h2>
          </div>
          <div
            class="col-auto"
          >
            <span
              class="tooltip-icon"
              data-testid="tooltip-icon-initiator-tooltip"
              id="initiator-tooltip"
            >
              <svg
                class="tooltip-icon"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 110c23.196 0 42 18.804 42 42s-18.804 42-42 42-42-18.804-42-42 18.804-42 42-42zm56 254c0 6.627-5.373 12-12 12h-88c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h12v-64h-12c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h64c6.627 0 12 5.373 12 12v100h12c6.627 0 12 5.373 12 12v24z"
                />
              </svg>
            </span>
          </div>
        </div>
        <div
          class="pb-2 row"
        >
          <div
            class="pr-0 text-left col-4"
          >
            <label
              class="c1"
            >
              Status
              :
            </label>
          </div>
          <div
            class="c2 text-left col"
          >
            <div
              class="form-group"
            >
              <select
                class="form-select form-control"
                id="input-documentStatusCode"
                name="documentStatusCode"
              >
                <option
                  class="option"
                  value="AMENDD"
                >
                  Amended
                </option>
                <option
                  class="option"
                  value="APPROVD"
                >
                  Approved
                </option>
                <option
                  class="option"
                  value="CNCLD"
                >
                  Cancelled
                </option>
                <option
                  class="option"
                  value="DRAFT"
                >
                  Draft
                </option>
                <option
                  class="option"
                  value="FINAL"
                >
                  Final
                </option>
                <option
                  class="option"
                  value="NONE"
                >
                  None
                </option>
                <option
                  class="option"
                  value="RGSTRD"
                >
                  Registered
                </option>
                <option
                  class="option"
                  value="SENT"
                >
                  Sent
                </option>
                <option
                  class="option"
                  value="SIGND"
                >
                  Signed
                </option>
                <option
                  class="option"
                  value="UNREGD"
                >
                  Unregistered
                </option>
              </select>
            </div>
          </div>
        </div>
        <h3
          class="c5"
        >
          Details
        </h3>
        <div
          class="c6 c7"
        >
          <div
            class="pb-2 row"
          >
            <div
              class="pr-0 text-left col-4"
            >
              <label
                class="c1"
              >
                Tag
                :
              </label>
            </div>
            <div
              class="c2 required text-left col"
            >
              <div
                class="required input form-group"
              >
                <input
                  class="form-control"
                  data-testid="metadata-input-Tag"
                  id="input-documentMetadata.1"
                  name="documentMetadata.1"
                  title=""
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            style="color: red;"
          />
        </div>
      </div>
      <div
        class="justify-content-end pt-4 row"
      >
        <div
          class="col-auto"
        >
          <button
            class="c8 Button btn btn-secondary"
            data-testid="cancel"
            type="button"
          >
            <div
              class="Button__value"
            >
              Cancel
            </div>
          </button>
        </div>
        <div
          class="col-auto"
        >
          <button
            class="c8 Button Button--disabled btn btn-primary"
            data-testid="save"
            disabled=""
            type="submit"
          >
            <div
              class="Button__value"
            >
              Save
            </div>
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;
